examples/us-treasury-auctions/TAuction.sol

```diff
@@ -0,0 +1,147 @@
+
+        if (rate == 0) {
+            // Store non-competitive bid
+            Suave.Bid memory bid = Suave.newBid(auctions[auctionId].startBlock, bidAllowedPeekers, bidAllowedStores, "tauction:v0:noncompetitive");
+            Suave.confidentialStore(bid.id, "tauction:v0:noncompetitive", confidentialInputs);
+            // Calculate current total and store for later
+            bytes memory totalData = Suave.confidentialRetrieve(bid.id, "tauction:v0:noncompetitivetotal");
+            uint256 currentTotal = totalData.length > 0 ? abi.decode(totalData, (uint256)) : 0;
+            uint256 newTotal = currentTotal + amount;
+            Suave.confidentialStore(bid.id, "tauction:v0:noncompetitivetotal", abi.encode(newTotal));
+            emit BidSubmitted(auctionId, bid.id);
+        } else {
+            // Store competitive bids, in order of rate offered
+            Suave.Bid memory bid = Suave.newBid(auctions[auctionId].startBlock, bidAllowedPeekers, bidAllowedStores, "tauction:v0:competitive");
+            Suave.Bid[] memory allCompetitiveSortedBids = Suave.fetchBids(blockHeight, "tauction:v0:competitiveSorted");
+
+            // TODO: I want to store competitive bids in ascending order, where 
+            // 1 is the lowest rate, 2 is slightly higher etc.
+            // overwriting ids won't work, and I am not sure how to add a new field to the confidentialInputs
+
+             // no sorted bids yet - goes in at rank 1
+            if (allCompetitiveSortedBids.length == 0) {
+                Suave.confidentialStore(1, "tauction:v0:competitiveSorted", confidentialInputs);
+            }
+            // some bids in the sorted array, need to insert new one
+            bool inserted = false;
+            uint i = 0;
+            
+            for (; i < allCompetitiveSortedBids.length; i++) {
+                // TODO: is this the correct way to retrieve the rate stored for any given bid?
+                uint16 rateToCompare = abi.decode(Suave.confidentialRetrieve(i, "tauction:v0:competitiveSorted"), (uint16));
+                
+                if (rate < rateToCompare) {
+                    // Insert the current bid here and shift the others
+                    shiftAndStoreBids(i, allCompetitiveSortedBids.length, confidentialInputs);
+                    Suave.confidentialStore(i, "tauction:v0:competitiveSorted", confidentialInputs);
+                    inserted = true;
+                    break;
+                }
+            }
+
+            if (!inserted) {
+                // If the current bid's rate is higher than all, insert at the end
+                Suave.confidentialStore(allCompetitiveSortedBids.length, "tauction:v0:competitiveSorted", confidentialInputs);
+            }
+
+            emit BidSubmitted(auctionId, bid.id);
+		}
+    }
+
+    // Complete an auction and emit results
+    // Anyone can call this, as the creator sets the duration
+    // question: can/should it emit all bids for public verification?
```

Comment:
It does not matter if you emit the bids, they are in the confidential store so they are not public. Normally, the pattern is to use the onchain callback to return the leak data of the bid that you want to expose.